Library/Homebrew/utils/pypi.rb

```diff
@@ -0,0 +1,116 @@
+# frozen_string_literal: true
+
+module PyPI
+  module_function
+
+  # Ensure pipgrip is installed
+  def ensure_pipgrip_installed
+    odie '"pipgrip" must be installed (`brew install pipgrip`)' unless Formula["pipgrip"].any_version_installed?
+  end
+
+  # Get name, url, and version for a given pypi package
+  def get_pypi_info(package, version)
+    metadata_url = "https://pypi.org/pypi/#{package}/#{version}/json"
+    out, _, status = curl_output metadata_url, "--location"
+
+    return unless status.success?
+
+    begin
+      json = JSON.parse out
+    rescue JSON::ParserError
+      return
+    end
+
+    sdist = json["urls"].find { |url| url["packagetype"] == "sdist" }
+    [json["info"]["name"], sdist["url"], sdist["digests"]["sha256"]]
+  end
+
+  def update_python_resources!(formula, version = nil, print_only: false, silent: false,
+                               ignore_non_pypi_packages: false)
+    # PyPI package name isn't always the same as the formula name. Try to infer from the URL.
+    pypi_name = if formula.stable.url.start_with?("https://files.pythonhosted.org/packages/")
+      File.basename(formula.stable.url).match(/^(.+)-[a-z\d.]+$/)[1]
+    else
+      formula.name
+    end
+
+    version ||= formula.version
+
+    if get_pypi_info(pypi_name, version).blank?
+      odie "\"#{pypi_name}\" at version #{version} is not available on PyPI." unless ignore_non_pypi_packages
+      return
+    end
+
+    non_pypi_resources = formula.resources.reject do |resource|
+      resource.url.start_with? "https://files.pythonhosted.org/packages/"
+    end
+
+    if non_pypi_resources.present? && !print_only
+      odie "\"#{formula.name}\" contains non-PyPI resources. Please update the resources manually."
+    end
+
+    ohai "Retrieving PyPI dependencies for \"#{pypi_name}==#{version}\"" if !print_only && !silent
+    pipgrip_output = Utils.popen_read Formula["pipgrip"].bin/"pipgrip", "--json", "#{pypi_name}==#{version}"
+    unless $CHILD_STATUS.success?
+      odie <<~EOS
+        Unable to determine dependencies for \"#{pypi_name}\" because of a failure when running
+        `pipgrip --json #{pypi_name}==#{version}`. Please update the resources for \"#{formula.name}\" manually.
+      EOS
+    end
+
+    packages = JSON.parse(pipgrip_output).sort.to_h
+
+    # Remove extra packages that may be included in pipgrip output
+    exclude_list = %W[#{pypi_name} argparse pip setuptools wsgiref]
```

Comment:
Should `wheel` be added here as well?

`pipgrip pipgrip` (not confusing at all) returns wheel but I think that `wheel` doesn't need to be listed as a resource.

Reply:
in conda, wheel is pre-installed, on pyenv it isn't. not sure about brew's python@3.8 though

Reply:
It looks like it is. In [`python@3.8`](https://github.com/Homebrew/homebrew-core/blob/43d95f9846b6d5b4fc52912d80173d827e760a6a/Formula/python%403.8.rb#L176-L178):

```ruby
    %w[setuptools pip wheel].each do |r|
      (libexec/r).install resource(r)
    end
```

Plus, `pip list` lists `wheel`